### DISCUSSION TITLE
Scoped middleware

### DISCUSSION THREAD

**Discussion by giladv** [trpc/trpc#4190] [General]
Hello,

Lets say we have an app that allows users to create organizations. each organization has it's own mapping of users to their privileges. 
All queries and mutations that are organization-scoped have a shared input `organizationId`. 
How one would create a permissions middleware type of a deal with trpc 10 current middleware setup?

i tried defining an input `organizationId` before the middleware but it's not available in the middleware itself. 

thanks!

**Comment by Nick-Lucas** [comment]
Middlewares can access the input ~in a type safe way~ (edit: strike the wrong statement) if the input precedes the middleware and IIRC multiple input calls get merged.

so you can add that ID as an initial input, then have a middleware which uses it, then stack your procedure calls on top with a more complete input

something like: 

publicProc.input.middleware.input.query

**Reply by giladv** [reply]
Hey, tried exactly that but the result input type inside the middleware is "unknown" for some reason. 
maybe a regression bug?

**Reply by Nick-Lucas** [reply]
Yep I see what you're saying, the middleware can't be aware of the inputs in a fully typesafe way because middlewares are created by `t.middleware` rather than in the `.input().use()` bit.

This should work though, and will safely type the client and procedure code, just the middleware has to check the expected key in a less safe way: https://stackblitz.com/edit/github-bfqein?file=server/index.ts

<img width="434" alt="image" src="https://user-images.githubusercontent.com/8896153/232253119-b1cf701f-57a4-475c-b463-74aeaa88567d.png">

**Comment by juliusmarminge** [comment]
Just remove the `t.middleware` and pass the middleware function directly in the `use` and it'll have all the necessary types

https://stackblitz.com/edit/github-bfqein-qe4c9w?file=server/index.ts

**Comment by giladv** [comment]
thanks!
